frontend/src/components/git-dialog.tsx

```diff
@@ -0,0 +1,241 @@
+  DialogDescription,
+  DialogHeader,
+  DialogTitle,
+} from "./ui/dialog";
+import { ScrollArea } from "./ui/scroll-area";
+import { Tabs, TabsContent, TabsList, TabsTrigger } from "./ui/tabs";
+
+interface CommitDetailsProps {
+  commit: Commit | null;
+}
+
+const CommitDetails = ({ commit }: CommitDetailsProps) => {
+  if (!commit) {
+    return (
+      <div className="text-muted-foreground text-sm text-center">
+        Select a commit to see its details
+      </div>
+    );
+  }
+  return (
+    <div className="space-y-2">
+      <div className="font-semibold items-center flex justify-between">
+        <div className="font-mono">
+          {commit.id.slice(0, 7)} {commit.summary}
+        </div>
+        <div>
+          <CopyButton value={commit.id} />
+        </div>
+      </div>
+      <div className="text-xs font-italic text-muted-foreground italic">
+        committed{" "}
+        {formatDistanceToNow(commit.time, {
+          addSuffix: true,
+        })}
+      </div>
+
+      <div className="prose prose-markdown max-w-none rounded-md py-2">
+        <Markdown children={commit.body} />
+      </div>
+      <div className="overflow-hidden text-sm text-muted-foreground font-mono">
+        <p>
+          <strong>Date:</strong> {new Date(commit.time).toLocaleString()}
+        </p>
+        <p>
+          <strong>Author:</strong> {commit.author.name}{" "}
+          {commit.author.email && <span> {commit.author.email}</span>}
+        </p>
+        <p>
+          <strong>Committer:</strong> {commit.committer.name}{" "}
+          {commit.author.email && <span> {commit.committer.email}</span>}
+        </p>
+      </div>
+    </div>
+  );
+};
+
+interface GitGraphData {
+  commits: Commit[];
+  diffs: Diff[];
+}
+
+const selector = (s: AppState) => ({
+  gitRevisions: s.gitRevisions,
+  clearGitRevisions: s.clearGitRevisions,
+});
+
+const uiSelector = (s: UiState) => ({
+  isOpen: s.isGitDialogOpen,
+  setIsOpen: s.setIsGitDialogOpen,
+});
+
+export const GitDialog = () => {
+  const { gitRevisions } = useStore(useShallow(selector));
+  const { isOpen, setIsOpen } = useUiStore(useShallow(uiSelector));
+
+  const [gitData, setGitData] = React.useState<GitGraphData>();
+  const [selectedCommit, setSelectedCommit] = React.useState<Commit | null>(
+    null,
+  );
+
+  const { theme, systemTheme } = useTheme();
+  const getTheme = (theme: string, systemTheme: string) => {
+    if (theme === "system") {
+      return systemTheme;
+    }
+    return theme;
+  };
+  const [isDarkMode, setIsDarkMode] = React.useState(
+    getTheme(theme ?? "system", systemTheme ?? "light") === "dark",
+  );
+  React.useEffect(() => {
+    setIsDarkMode(
+      getTheme(theme ?? "system", systemTheme ?? "light") === "dark",
+    );
+  }, [theme, systemTheme]);
+
+  React.useEffect(() => {
+    // This effect can be used to perform any side effects when the panel opens or closes
+    if (isOpen && gitRevisions.length >= 1) {
+      client
```

Comment:
please add a helper function to `client.ts` for fetching the commits. Then all the backend API request are bundled in one place and it is easier to update them if needed.

Then the mapping of the error message is also encapsulated.